File: py/shuffleArray.py
# 384.Shuffle An Array:设计一个数组结构，可以返回随机打乱的数组，或返回原始数组
#另见 java实现
import random
class Solution(object):
    def __init__(self, nums): #33.56%，算hack解法
        self.reset = lambda: nums
        self.shuffle = lambda: random.sample(nums, len(nums))
#numpy.random.sample(size=None) 返回半开区间的[0.0, size)的浮点数
#random.sample(population, k) 从population中随机选出k个数
s = Solution([1, 2, 3, 4])
print s.reset()
print s.shuffle()
print s.reset()

File: march2018/ShuffleArray.java
package march2018;

import java.util.Random;

public class ShuffleArray {

}

/**58.4%*/
class Solution {
    private int[] nums;

    public Solution(int[] nums) {
        this.nums = nums;
    }

    /** Resets the array to its original configuration and return it. */
    public int[] reset() {
        return nums;
    }

    /** Returns a random shuffling of the array. */
    public int[] shuffle() {
        int[] dizz = new int[nums.length];
        for (int i = 0; i < nums.length; i++) {//new Random().nextInt(bound); [0, bound)
            int r = (int) (Math.random() * (i + 1));//random 0.0 ~ 1.0 -> r=0 ~ i
            dizz[i] = dizz[r];//每次一定会把前面的数字和后边的换
            dizz[r] = nums[i];
        }

        return dizz;
    }
}
